File: src/test/java/com/fluxchess/pulse/BitboardTest.java
/*
 * Copyright (C) 2013-2019 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */
package com.fluxchess.pulse;

import com.fluxchess.pulse.model.Square;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.LinkedList;
import java.util.Random;

import static com.fluxchess.pulse.Bitboard.add;
import static com.fluxchess.pulse.model.Square.a6;
import static org.assertj.core.api.Assertions.assertThat;

class BitboardTest {

	private LinkedList<Integer> pool = null;

	@BeforeEach
	void setUp() {
		Random random = new Random();
		pool = new LinkedList<>();

		while (pool.size() < Long.SIZE) {
			int value = random.nextInt(Long.SIZE);
			if (!pool.contains(Square.values[value])) {
				pool.add(Square.values[value]);
			}
		}
	}

	@Test
	void shouldAddAllSquaresCorrectly() {
		long bitboard = 0;

		for (int x88square : pool) {
			bitboard = add(x88square, bitboard);
		}

		assertThat(bitboard).isEqualTo(-1L);
	}

	@Test
	void shouldRemoveAllSquaresCorrectly() {
		long bitboard = -1;

		for (int x88square : pool) {
			bitboard = Bitboard.remove(x88square, bitboard);
		}

		assertThat(bitboard).isEqualTo(0L);
	}

	@Test
	void shouldReturnTheNextSquare() {
		long bitboard = add(a6, 0);

		int square = Bitboard.next(bitboard);

		assertThat(square).isEqualTo(a6);
	}

	@Test
	void shouldReturnCorrectRemainder() {
		long bitboard = 0b1110100;

		long remainder = Bitboard.remainder(bitboard);

		assertThat(remainder).isEqualTo(0b1110000L);
	}

	@Test
	void shouldReturnCorrectSize() {
		long bitboard = 0b111;

		int size = Bitboard.size(bitboard);

		assertThat(size).isEqualTo(3);
	}
}

File: src/test/cpp/bitboardtest.cpp
/*
 * Copyright (C) 2013-2019 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

#include "bitboard.h"
#include "model/square.h"

#include "gtest/gtest.h"

#include <random>
#include <list>

using namespace pulse;

class BitboardTest : public ::testing::Test {
protected:
	std::list<int> pool;

	virtual void SetUp() {
		std::default_random_engine generator;

		while (pool.size() < 64) {
			std::uniform_int_distribution<int> distribution(0, 63);
			int value = distribution(generator);
			if (std::find(pool.begin(), pool.end(), square::values[value]) == pool.end()) {
				pool.push_back(square::values[value]);
			}
		}
	}
};

TEST_F(BitboardTest, shouldAddAllSquaresCorrectly) {
	uint64_t bitboard = 0;

	for (auto x88square: pool) {
		bitboard = bitboard::add(x88square, bitboard);
	}

	EXPECT_EQ(bitboard, std::numeric_limits<uint64_t>::max());
}

TEST_F(BitboardTest, shouldRemoveAllSquaresCorrectly) {
	uint64_t bitboard = std::numeric_limits<uint64_t>::max();

	for (auto x88square: pool) {
		bitboard = bitboard::remove(x88square, bitboard);
	}

	EXPECT_EQ(bitboard, 0);
}

TEST(bitboardtest, shouldReturnTheNextSquare) {
	uint64_t bitboard = bitboard::add(square::a6, 0);

	int square = bitboard::next(bitboard);

	EXPECT_EQ(square, +square::a6);
}

TEST(bitboardtest, shouldReturnCorrectRemainder) {
	uint64_t bitboard = 0b1110100;

	uint64_t remainder = bitboard::remainder(bitboard);

	EXPECT_EQ(remainder, 0b1110000);
}

TEST(bitboardtest, shouldReturnCorrectSize) {
	uint64_t bitboard = 0b111;

	int size = bitboard::size(bitboard);

	EXPECT_EQ(size, 3);
}

TEST(bitboardtest, testNumberOfTrailingZeros) {
	uint64_t bitboard = 0;
	int i = 0;

	for (auto square: square::values) {
		bitboard = bitboard::add(square, bitboard);

		EXPECT_EQ(i, bitboard::numberOfTrailingZeros(bitboard));

		bitboard = bitboard::remove(square, bitboard);
		i++;
	}
}

TEST(bitboardtest, testBitCount) {
	std::default_random_engine generator;

	for (int i = 0; i < 1000; i++) {
		uint64_t bitboard = 0;
		int count = 0;

		int index = 0;
		while (true) {
			std::uniform_int_distribution<int> distribution(1, 4);
			index += distribution(generator);
			if (index < 64) {
				bitboard |= 1ULL << index;
				count++;
			} else {
				break;
			}
		}

		EXPECT_EQ(count, bitboard::bitCount(bitboard));
	}
}
